### DISCUSSION TITLE
Connections management during MikroORM migration

### DISCUSSION THREAD

**Discussion by evenfrost** [mikro-orm/mikro-orm#2457] [General]
Recently, when deploying a database update to a remote environment, we've run an arbitrary migration via `mikro-orm migration:up`, and this caused a spike of connections over the limit that resulted in the `remaining connection slots are reserved for non-replication superuser connections` error, and our API becoming completely unresponsive for the period of migration. I've had to cancel the migration as it was taking a long time (which was not the issue itself but frozen API).

I'm wondering since I haven't found this info in the documentation, how the connections are managed during the migration and how to control them?  We have about 40-50 active connections usually, with the cap of 396 connections (Postgres on DigitalOcean Managed Databases), and claiming all free connections for the single migration doesn't quite fit in the process.

Here is the migration that was run:

```ts
import { Migration } from '@mikro-orm/migrations';

export class Migration20211022164646 extends Migration {
  async up(): Promise<void> {
    this.addSql('alter table "productBrands" drop constraint "productBrands_clientId_fkey"');
    this.addSql('alter table "productBrands" add constraint "productBrands_clientId_fkey" foreign key ("clientId") references "clients" ("id") on update cascade on delete cascade;');
    this.addSql('delete from "productBrands" where "clientId" is null');

    this.addSql('alter table "productDynamicContentItems" drop constraint "productDynamicContentItems_clientId_fkey"');
    this.addSql('alter table "productDynamicContentItems" add constraint "productDynamicContentItems_clientId_fkey" foreign key ("clientId") references "clients" ("id") on update cascade on delete cascade;');
    this.addSql('delete from "productDynamicContentItems" where "clientId" is null');

    this.addSql('alter table "productGroups" drop constraint "productGroups_clientId_fkey"');
    this.addSql('alter table "productGroups" add constraint "productGroups_clientId_fkey" foreign key ("clientId") references "clients" ("id") on update cascade on delete cascade;');
    this.addSql('delete from "productGroups" where "clientId" is null');

    this.addSql('alter table "productUsers" drop constraint "productUsers_clientId_fkey"');
    this.addSql('alter table "productUsers" add constraint "productUsers_clientId_fkey" foreign key ("clientId") references "clients" ("id") on update cascade on delete cascade;');
    this.addSql('delete from "productUsers" where "clientId" is null');
  }

  async down(): Promise<void> {
    this.addSql('alter table "productBrands" drop constraint "productBrands_clientId_fkey"');
    this.addSql('alter table "productBrands" add constraint "productBrands_clientId_fkey" foreign key ("clientId") references "clients" ("id") on update cascade on delete set null;');

    this.addSql('alter table "productDynamicContentItems" drop constraint "productDynamicContentItems_clientId_fkey"');
    this.addSql('alter table "productDynamicContentItems" add constraint "productDynamicContentItems_clientId_fkey" foreign key ("clientId") references "clients" ("id") on update cascade on delete set null;');

    this.addSql('alter table "productGroups" drop constraint "productGroups_clientId_fkey"');
    this.addSql('alter table "productGroups" add constraint "productGroups_clientId_fkey" foreign key ("clientId") references "clients" ("id") on update cascade on delete set null;');

    this.addSql('alter table "productUsers" drop constraint "productUsers_clientId_fkey"');
    this.addSql('alter table "productUsers" add constraint "productUsers_clientId_fkey" foreign key ("clientId") references "clients" ("id") on update cascade on delete set null;');
  }
}
```

**Comment by B4nan** [comment]
That sounds quite weird, I would say only one connection should be used for migrations. Under the hood its knex that manages connections and umzug that is responsible for migrations, it's not really managed in MikroORM itself. And for migrations we set the pool size to a single connection to enforce this.

https://github.com/mikro-orm/mikro-orm/blob/master/packages/cli/src/commands/MigrationCommandFactory.ts#L87

**Comment by evenfrost** [comment]
I've done some more testing (as much as I can do on the production environment), and can see this particular query in the migration file causes connections number to grow up indefinitely till the connection limit is reached:

```ts
this.addSql('delete from "productBrands" where "clientId" is null');
```

This is a long-running query as `productBrand` entities have relations in other tables.
Here is how it looks like in `pgcli`, when running `SELECT count(*) FROM pg_stat_activity; \watch 2`:

![Screenshot from 2021-11-24 11-39-52](https://user-images.githubusercontent.com/3055750/143216223-d18d1e09-9f4a-42d7-aec4-ae1fe463aeb3.png)

The number of connections starts growing more and more, and during this our DB-related backend infrastructure becomes non-responsive as well.

This is how it looks when I manually cancel this query and force end the migration:

![Screenshot from 2021-11-24 11-38-16](https://user-images.githubusercontent.com/3055750/143217017-5842d850-e89f-44a6-8981-a03c5f90e85a.png)

The number of connections instantly goes back to normal (~60), and the database starts being responsive again.

When I run this query manually in `pgcli`:
```sql
delete from "productBrands" where "clientId" is null
```

There is no spike of connections (though the query itself takes a long time to complete anyway, which is not a problem):

![Screenshot from 2021-11-24 11-41-04](https://user-images.githubusercontent.com/3055750/143217564-4348ba52-ed3e-44ac-a8a5-bd91f9aa5591.png)

This may look like some internal loop trying to acquire more and more connections on each tick, but may be something else as well.
It's not critical migration, and we can either skip it or rewrite to utilize Sequelize migrations, which we still use for the majority of our DB ORM logic, but it is still good to understand where it's coming from as future migrations can cause this as well.

**Reply by B4nan** [reply]
Would be good if you could try to reproduce this only with knex, as I really don't think it's something at our end. Are you using v4 or v5? The knex version in v4 is a bit old as they introduced BCs (TS 4.1 support) in 0.95, so could be as well something already fixed and available in v5 and latest knex.

**Reply by evenfrost** [reply]
I'm using `v4.5.9`, so it may be better in my case to just wait for the stable `v5` release with the updated knex.

I was a bit confused to see knex  `v0.21.x` in my `package-lock.json` considering the fact their latest is `v0.95.x` but then realized there were many breaking changes for that version, including some improvements for transactions.  So maybe it's already fixed somewhere.

Anyway, thanks for looking into it, and keep up the great work. 💪🏻
